feat: Add EKS Fargate support

itssimon opened the pull request:
# PR o'clock

## Description

1) Add ingress rule to workers and cluster security groups to allow Fargate pods to communicate with EC2 nodes
2) Allow EKS Fargate profiles to be created via input `eks_fargate_profiles = [...]`. This also creates an IAM role for pod execution and extends the aws-auth ConfigMap.

### Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

barryib commented:
@itssimon Thanks for working on this. But since this module is getting more complex and discussed in https://github.com/terraform-aws-modules/terraform-aws-eks/issues/635 or https://github.com/terraform-aws-modules/terraform-aws-eks/issues/774. We decided to split functionality into different submodules. We started recently with the [managed node groups](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/650).

So my point here is this feature should be added as a submodule.

Can you please work in that direction ?

itssimon commented:
Done @barryib

itssimon commented:
Is there anything else required for this to get approval and be merged?

barryib commented on aws_auth.tf:
```diff
@@ -38,22 +38,31 @@ locals {
     }
   ]
 
+  auth_fargate_roles = [
+    for index in range(0, var.create_eks && var.create_eks_fargate ? 1 : 0) : {
+      worker_role_arn = module.fargate.iam_role_arn
+      platform        = "fargate"
+    }
+  ]
+
```
Add this as output in the fargate submodule like https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/modules/node_groups/outputs.tf#L6-L14

barryib commented:
> Is there anything else required for this to be approved and merged?

Sorry for the late answer. I was busy these last days. Reviewed.

Thanks again for working on this.

barryib commented on variables.tf:
```diff
@@ -332,3 +332,18 @@ variable "cluster_encryption_config" {
+
+variable "eks_fargate_profiles" {
+  description = "EKS Fargate profiles"
+  type = list(object({
```
Use map of object please instead a list. More info here https://github.com/terraform-aws-modules/terraform-aws-eks/issues/476. And we are trying to move from count to for_each in all ressources.

itssimon commented on modules/fargate/fargate.tf:
```diff
@@ -0,0 +1,67 @@
+  protocol                 = "all"
+  security_group_id        = var.cluster_primary_security_group_id
+  source_security_group_id = var.worker_security_group_id
+}
```
See https://docs.aws.amazon.com/eks/latest/userguide/fargate-getting-started.html

> If you are working with an existing cluster that already has worker nodes associated with it, you need to make sure that pods on these nodes can communicate freely with pods running on Fargate. Pods running on Fargate are automatically configured to use the cluster security group for the cluster that they are associated with. You must ensure that any existing worker nodes in your cluster can send and receive traffic to and from the cluster security group.

barryib replied:
From my understanding , these are already created by the root module (https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/workers.tf#L297-L350) and https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/cluster.tf#L101-L110

Maybe I'm getting something wrong ?

dpiddockcmp replied:
This block of code sets up the security group permissions between the default security group created by AWS and the worker groups the module creates.

This block should be moved into the main module, in a new PR, to fix https://github.com/terraform-aws-modules/terraform-aws-eks/issues/877#issuecomment-631116579 . Not sure if #877 itself was caused by this. 

Potentially breaking change as it may conflict with rules that users running a mix of "classic" workers and managed node groups or fargate could already have put in.

itssimon replied:
I can confirm that, without these additional rules, Fargate pods are unable to communicate with pods running on "classic" worker group nodes. This does not affect managed node groups.

barryib replied:
arf... I see. I'm agree with @dpiddockcmp this should be moved in the root module.

@itssimon can you please do that in a separate PR ? We'll hold this until we merge that PR.

> Potentially breaking change as it may conflict with rules that users running a mix of "classic" workers and managed node groups or fargate could already have put in.

@dpiddockcmp by conflict, you mean it'll add a new rule which probably duplicates the rule added by users ?

dpiddockcmp replied:
AWS security groups do not allow duplicated rules. The API returns an error. Only impacts users running mixed clusters who have implemented the rules exactly as above. We have no idea how many that might be.

itssimon replied:
Done. See https://github.com/terraform-aws-modules/terraform-aws-eks/pull/892.

itssimon commented:
Thanks @barryib. I addressed all your review notes.

barryib commented on modules/fargate/variables.tf:
```diff
@@ -0,0 +1,39 @@
+variable "create" {
```
I was thinking about using [`create_eks`](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/580#issuecomment-554312341) as variable name for consistency => it's a naming convention.

itssimon replied:
I've actually removed the `create` variable entirely, as it wasn't really necessary. Instead it is sufficient to check if `profiles` is empty or not.

barryib commented on fargate.tf:
```diff
@@ -0,0 +1,9 @@
+module "fargate" {
+  source                            = "./modules/fargate"
+  cluster_name                      = aws_eks_cluster.this[0].name
```
```hcl
cluster_name = coalescelist(aws_eks_cluster.this[*].name, [""])[0]
```
Use splat syntax please to avoid this kind of problem https://github.com/terraform-aws-modules/terraform-aws-eks/pull/868

itssimon replied:
Done

barryib commented on fargate.tf:
```diff
@@ -0,0 +1,9 @@
+module "fargate" {
+  source                            = "./modules/fargate"
+  cluster_name                      = aws_eks_cluster.this[0].name
+  profiles                          = var.create_eks ? var.eks_fargate_profiles : {}
```
```suggestion
  profiles   = var.eks_fargate_profiles
  create_eks = var.create_eks
```

For consistency I explained at https://github.com/terraform-aws-modules/terraform-aws-eks/pull/866#discussion_r427949268

itssimon replied:
I've removed the `create_eks_fargate` variable based on your other comment, and I think it made perfect sense to just check if any profiles are defined. The additional variable added no value and I don't think we should add it as `create_eks` again. The `create_eks` variable in the root module is still being honored by setting profiles = {} if it's false.

barryib replied:
You're absolutely right, you can compute it with the `length` function. But, my point here is more about coding style which should match as much as possible with the rest of the module. So :

- it's more like a coding style consistency with the root and the first submodule
- it improve readability. No need for someone who's reading your code to try to compute it in his head. `create_eks` is `false` so it won't create any ressources, period.
- you don't need to compute it twice (even if it's a small thing here) when you're calling module and inside the module. You just pass the `create_eks` down to your resources

So I'll definitely prefer this writing in that context

```hcl
module "fargate" {
  source                            = "./modules/fargate"
  cluster_name                      = coalescelist(aws_eks_cluster.this[*].name, [""])[0]
  create_eks                        = var.create_eks
  fargate_profiles                  = var.fargate_profiles
  subnets                           = var.subnets
  tags                              = var.tags
  cluster_primary_security_group_id = element(concat(aws_eks_cluster.this[*].vpc_config[0].cluster_security_group_id, list("")), 0)
  worker_security_group_id          = local.worker_security_group_id
}
```

barryib commented on variables.tf:
```diff
@@ -332,3 +332,12 @@ variable "cluster_encryption_config" {
+  description = "EKS Fargate profiles"
+  type = map(object({
+    namespace = string
+    labels    = map(string)
```
would we need different tags by profile ? If yes, we can probably add a `additional_tags` like what we already have in node group submodule https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/modules/node_groups/node_groups.tf#L40-L43

itssimon replied:
I don't see why that would be necessary.

barryib replied:
I don't use fargate, but you mean that no one would want to do something like 
```hcl
eks_fargate_profiles = {
  profile-1 = {
    namespace = "namespace-1"
    labels = {
      app = "foo"
    },
    additional_tags = {
      ownner = "team-foo"
    }
  },
  profile-2 = {
    namespace = "namespace-2"
    labels = {
      app = "bar"
    },
    additional_tags = {
      ownner = "team-bar"
    }
  }
}

}
```

dalgibbard commented:
Hey all,
Apologies if this muddies the waters, but i've been trying to test this with keenness to implement it in our deployments :)

Can you confirm that this does allow for fargate-only cluster support too?
For example; i'm running with the following config:

```
data "aws_eks_cluster" "cluster" {
  name = module.eks_cluster.cluster_id
}

data "aws_eks_cluster_auth" "cluster" {
  name = module.eks_cluster.cluster_id
}

resource "aws_kms_key" "eks_cluster_key" {
  description = "KMS Key for use by EKS Fargate"
  tags        = var.tags
  key_usage   = "ENCRYPT_DECRYPT"
}

resource "aws_kms_alias" "eks_cluster_key_alias" {
  name = "alias/${var.eks_cluster_name}_eks_key"
  target_key_id = aws_kms_key.eks_cluster_key.id
}

resource "aws_kms_key" "eks_log_group_key" {
  description = "KMS Key for use for encrypting EKS logs"
  tags        = var.tags
  key_usage   = "ENCRYPT_DECRYPT"
  policy = jsonencode(
    {
      Version = "2012-10-17",
      Statement = [{
        Effect = "Allow"
        Principal = {
          AWS = "arn:aws:iam::${var.aws_account_id}:root"
        }
        Action = [
          "kms:*"
          ]
        Resource = "*"
      },
      {
        Effect = "Allow"
        Principal = {
          Service = "logs.${var.aws_region}.amazonaws.com"
        }
        Action = [ 
          "kms:Encrypt*",
          "kms:Decrypt*",
          "kms:ReEncrypt*",
          "kms:GenerateDataKey*",
          "kms:Describe*"
        ]
        Resource = "*"
      }]
  })
}

resource "aws_kms_alias" "eks_log_group_alias" {
  name = "alias/${var.eks_cluster_name}_eks_log_group_key"
  target_key_id = aws_kms_key.eks_log_group_key.id
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.cluster.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
  token                  = data.aws_eks_cluster_auth.cluster.token
  load_config_file       = false
  version                = "~> 1.9"
}

module "eks_cluster" {
    // TODO: point this to the main repo, once PR-886 is merged.
    //  https://github.com/terraform-aws-modules/terraform-aws-eks/pull/866
  source          = "git::https://github.com/terraform-aws-modules//terraform-aws-eks.git?ref=fargate"
  cluster_name    = var.eks_cluster_name
  cluster_version = var.eks_cluster_version
  subnets         = var.subnet_ids_list
  vpc_id          = var.vpc_id
  tags            = var.tags
  eks_fargate_profiles = {
      {
          namespace = "namespace1",
          labels = {}
      },
      {
          namespace = "namespace2",
          labels = {}
      }
  }

  worker_groups = []
  cluster_enabled_log_types = var.cluster_enabled_log_types
  manage_aws_auth = true
  map_roles = var.map_roles
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access = false
  create_eks = true
  cluster_encryption_config = [
      {
          provider_key_arn = aws_kms_key.eks_cluster_key.arn
          resources = ["secrets"]
      }
  ]
  cluster_log_retention_in_days = var.cluster_log_retention_in_days
  cluster_log_kms_id = aws_kms_key.eks_log_group_key.id
}
```

With this, it's creating an EKS cluster (w/security groups, IAM roles etc), but not any of the aws-auth or fargate policies etc. Am I missing something?

barryib commented:
@dalgibbard This module can't be used (for now) for fargate or managed node group only. We just started split the terraform-aws-eks module into different submodule. Reasons have discussed in https://github.com/terraform-aws-modules/terraform-aws-eks/issues/635 and https://github.com/terraform-aws-modules/terraform-aws-eks/issues/774.

So for now, we don't split the `aws-auth` into a submodule yet. We have it in mind, but at this I don't see any PR in that direction.

dalgibbard commented:
> @dalgibbard This module can't be used (for now) for fargate or managed node group only. We just started split the terraform-aws-eks module into different submodule. Reasons have discussed in #635 and #774.
> 
> So for now, we don't split the `aws-auth` into a submodule yet. We have it in mind, but at this I don't see any PR in that direction.

Just to confirm, I'm using this PR content, and then calling the main module with the correct params; I'm just defining any node groups/workers as I want only Fargate instances. Should that work?

barryib commented:
I don't know. You can probably play with node group and its desired capacity.

But if you want those kind configuration you can probably try https://github.com/cloudposse/terraform-aws-eks-fargate-profile/blob/master/README.md which offer more flexibility, but is less straightforward for new users.

dalgibbard commented:
That's kinda the point when running solely on EKS Fargate, there is no node group or workers :) the repo you linked is just the Fargate profiles from what I can tell, so not too handy. This PR is looking ideal to provision a cluster for EKS Fargate use anyway. I'll wait and see if @itssimon has any thoughts; else I'll just wait for the merge and test after :+1: 
It's probably just me doing something dumb, I just wanted to see if this had been used for that purpose yet was all, and if there were any working examples :)

barryib commented:
My point was putting the desired capacity to 0. But I personally won't do that in production because, it is hacky.

The repo I was referring took the decision to split eks into different module (not submodule) with their own repo. So in the readme you can see an example on how to spinup an EKS cluster and fargate profiles.

You can see the module source reference these two repos:

- https://github.com/cloudposse/terraform-aws-eks-cluster
- https://github.com/cloudposse/terraform-aws-eks-fargate-profile

dalgibbard commented:
Another possible 'gotcha' here, and I haven't seen it explicitly handled in this PR just yet (apologies if i missed it)-
Creating multiple fargate profiles can trip the following error:
```
Error: error creating EKS Fargate Profile (my_cluster:eks_fargate_profile_myprofile2): ResourceInUseException: Cannot create Fargate Profile eks_fargate_profile_myprofile2 because cluster my_cluster currently has Fargate profile eks_fargate_profile_myprofile1 in status CREATING
```
So need to wait for profile completion one-by-one before trying to create another :/

itssimon commented:
@dalgibbard This PR was not aimed at supporting Fargate-only clusters. Correct me if I'm wrong, but given all the limitations that come with running pods on Fargate (no persistent storage, no DaemonSets, no bursting, etc.) I don't think Fargate-only clusters are actually viable for most use cases yet.

That one-by-one creation of profiles due to the ResourceInUseException is indeed a problem. Can anyone think of a solution for this?

itssimon commented on modules/fargate/fargate.tf:
```diff
@@ -0,0 +1,67 @@
+resource "aws_eks_fargate_profile" "this" {
+  for_each               = var.profiles
+  cluster_name           = var.cluster_name
+  fargate_profile_name   = format("%s-fargate-%s", var.cluster_name, each.value.namespace)
```
Great idea, I've added that. But when I tested it, the name was never actually passed down to the module because the object type definition (in `variables.tf`) doesn't specify `name` as a key. I don't want `name` to be a required argument, but an optional argument is [not supported yet](https://github.com/hashicorp/terraform/issues/19898).

A workaround would be to not use the object type definition at all, but then we wouldn't get the automatic type check and the implicit documentation of allowed keys and their types.

dalgibbard commented:
> @dalgibbard This PR was not aimed at supporting Fargate-only clusters. Correct me if I'm wrong, but given all the limitations that come with running pods on Fargate (no persistent storage, no DaemonSets, no bursting, etc.) I don't think Fargate-only clusters are actually viable for most use cases yet.
> 
> That one-by-one creation of profiles due to the ResourceInUseException is indeed a problem. Can anyone think of a solution for this? How to achieve that is beyond my knowledge of Terraform.

Actually in hindsight, the terraform creation of these resources is blocking, so they should be created one by one in terraform already; I was running terragrunt which parallelized this step and that's what triggered it. Should be ignorable I think.

itssimon commented:
I think I addressed all review notes.
Also removed the SG rules from this PR and a created separate PR for them: https://github.com/terraform-aws-modules/terraform-aws-eks/pull/892.

dpiddockcmp commented on modules/fargate/outputs.tf:
```diff
@@ -0,0 +1,19 @@
+
+output "aws_auth_roles" {
+  description = "Roles for use in aws-auth ConfigMap"
+  value = [
+    for index in range(0, var.create_eks ? 1 : 0) : {
+      worker_role_arn = element(concat(aws_iam_role.eks_fargate_pod.*.arn, list("")), 0)
+      platform        = "fargate"
+    }
+  ]
```
This doesn't need to be this complicated.
```hcl
  value = [
    for role in aws_iam_role.eks_fargate_pod : {
      worker_role_arn = role.arn
      platform        = "fargate"
    }
  ]
```

dpiddockcmp commented on modules/fargate/variables.tf:
```diff
@@ -0,0 +1,39 @@
+  type        = string
+}
+
+variable "worker_security_group_id" {
```
I don't think this is required any more

dpiddockcmp commented on modules/fargate/variables.tf:
```diff
@@ -0,0 +1,39 @@
+  type        = map(string)
+}
+
+variable "cluster_primary_security_group_id" {
```
I don't think this is required any more

dpiddockcmp commented on modules/fargate/fargate.tf:
```diff
@@ -0,0 +1,44 @@
+
+resource "aws_iam_role" "eks_fargate_pod" {
+  count              = var.create_eks ? 1 : 0
+  name               = format("%s-fargate", var.cluster_name)
```
The cluster and worker roles are created using `name_prefix` by default. This allows for the same name cluster to be created in different regions as IAM is global. I have no idea if anybody actually does that. There is also the complexity around `name_prefix` vs `name` and `var.workers_role_name`. Do we want to replicate that here? @barryib

itssimon replied:
This is the only open topic to resolve now. How would you like to handle the naming here @barryib ?

itssimon commented on fargate.tf:
```diff
@@ -0,0 +1,10 @@
+module "fargate" {
```
Could you provide some guidance on how to best achieve that?

dpiddockcmp replied:
[`node_groups.tf`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/v12.0.0/node_groups.tf) in the root of the module has one example. But honestly I think the rejected work in #867 is a more "terraform native" way of implementing it. Makes the module easier to use stand-alone as well. My mistake as I didn't think depending on vars had been implemented yet when I moved managed node groups to a module.

itssimon replied:
I've implemented this like in `node_groups.tf` for consistency.

itssimon replied:
Creating this dependency creates a cycle. Are you sure the fargate module would cause the aws-auth ConfigMap to be created?

```
Error: Cycle: data.null_data_source.fargate, module.fargate.var.cluster_name, module.fargate.aws_iam_role.eks_fargate_pod, module.fargate.output.aws_auth_roles, local.configmap_roles, kubernetes_config_map.aws_auth
```

itssimon commented on modules/fargate/fargate.tf:
```diff
@@ -0,0 +1,44 @@
+# EKS Fargate Pod Execution Role
+
+data "aws_iam_policy_document" "eks_fargate_pod_assume_role" {
```
This rule is based on instructions on how to set up EKS Fargate in the [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/fargate-getting-started.html). I don't think you can make Fargate work with EKS without it.

dpiddockcmp replied:
Yes, you need an IAM role. But people working in very restrictive environments will have IAM roles created by a central trusted team and then developers can use them. Or permissions boundaries. See #266 and #282

itssimon replied:
Okay, I added a variable `create_fargate_pod_execution_role` and made it `false` by default so that the role doesn't get created unless people actually want to used Fargate with their EKS cluster.

dpiddockcmp commented on modules/fargate/fargate.tf:
```diff
@@ -0,0 +1,44 @@
+  }
+}
+
+resource "aws_iam_role" "eks_fargate_pod" {
```
Can you add in the `iam_path` variable from the parent too

dpiddockcmp commented on variables.tf:
```diff
@@ -338,3 +338,18 @@ variable "cluster_encryption_config" {
+variable "create_fargate_pod_execution_role" {
+  description = "Controls if the EKS Fargate pod execution IAM role should be created."
+  type        = bool
+  default     = false
```
Could creation of the IAM role be based on the length of `fargate_profiles`? That way this variable could default to true.

And if we're giving the option of not creating the fargate role we should provide the ability to pass in an externally created role ARN.

dpiddockcmp commented:
A vote was kind of taken on #774 where a majority supported going full in on TF 0.13's for_each module support instead of looping inside the module. But 0.13 isn't exactly working too well with the module currently. 

Raises design questions for this PR as the module contains an IAM role.

barryib commented:
Thanks @itssimon  for working on this. We have an terraform-aws-modules working session this friday. We'll discuss about the direction we want to take about this feature. We'll come back to you pretty soon.

barryib commented:
I'm closing this PR in favor of #1067. That PR takes all the great work of @itssimon and rework it to add requested changes.
